libs/@local/hash-graph-types/graph_types/_const.py

```diff
@@ -0,0 +1,70 @@
+from typing import Annotated, Any, TypeVar, cast, overload, Literal
+
+from pydantic import GetCoreSchemaHandler, GetJsonSchemaHandler
+from pydantic.json_schema import JsonSchemaValue
+from pydantic_core import CoreSchema, core_schema
+
+T = TypeVar("T")
+
+
+class ConstAnnotation:
+    const: Any
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls,
+        source_type: Any,  # noqa: ANN401
+        handler: GetCoreSchemaHandler,
+    ) -> CoreSchema:
+        schema = handler(source_type)
+
+        return core_schema.no_info_after_validator_function(
+            cls.validate_const,
+            schema,
+        )
+
+    @classmethod
+    def __get_pydantic_json_schema__(
+        cls,
+        schema: CoreSchema,
+        handler: GetJsonSchemaHandler,
+    ) -> JsonSchemaValue:
+        json_schema = handler(schema)
+        json_schema.update(const=cls.const)
+
+        return json_schema
+
+    @classmethod
+    def validate_const(cls, v: T) -> T:
+        if v != cls.const:
+            msg = f"Value must be {cls.const}"
+            raise ValueError(msg)
+
+        return v
+
+
+_Undefined = object()
+
+
+@overload
+def constant(const_: T, /) -> type[T]:
+    ...
+
+
+@overload
+def constant(type_: type[T], const_: T, /) -> type[T]:
+    ...
+
+
+def constant(type_: type[T] | T, const_: T | object = _Undefined, /) -> type[T]:
```

Comment:
I don't really understand this signature. Why `type_` can be `type[T]` **or** `T`? Why `const_` is allowed to be `object`?
What was wrong with the previous approach - what is the difference?

Reply:
Python has no native support for sentinel values; this typedef is not what the end user will see, only the overloaded one. The `| object` is to allow for the sentinel value `_Undefined`. Otherwise we would need to add a default value to `None` or similar and that could lead to clashes and false positives.

Reply:
The previous approach has nothing wrong with it (although the validate function lacked a `return`) meaning no value was returned after validation.

Reply:
I took the issue advice and then transformed it into a function to make it more composable in the future (and make it clear what's happening)